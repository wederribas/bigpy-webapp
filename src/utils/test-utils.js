import React from 'react'
import { render } from 'react-testing-library'
import { MuiThemeProvider, createMuiTheme } from '@material-ui/core/styles'

const theme = createMuiTheme({
  typography: {
    useNextVariants: true
  }
})

const customRender = node => {
  return render(<MuiThemeProvider theme={theme}>{node}</MuiThemeProvider>)
}

export * from 'react-testing-library'

export { customRender as render }
