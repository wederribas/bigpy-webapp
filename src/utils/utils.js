// @flow

/**
 * Deburr special Latin letters to basic Latin letters and removing
 * combining diacritical marks.
 */
export function deburr(string: string): string {
  return string.normalize('NFD').replace(/[\u0300-\u036f]/g, '')
}
