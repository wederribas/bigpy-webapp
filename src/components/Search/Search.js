import React, { Component } from 'react'
import { withRouter } from 'react-router'
import PropTypes from 'prop-types'
import Downshift from 'downshift'
import { graphql } from 'react-apollo'
import gql from 'graphql-tag'
import { withStyles } from '@material-ui/core/styles'
import Input from '@material-ui/core/Input'
import Paper from '@material-ui/core/Paper'
import MenuItem from '@material-ui/core/MenuItem'
import MoodBad from '@material-ui/icons/MoodBad'
import Grid from '@material-ui/core/Grid'
import Typography from '@material-ui/core/Typography'
import BouncingLoader from '../BouncingLoader/BouncingLoader'
import { deburr } from '../../utils/utils'

const MAX_SUGGESTION_ITEMS = 8

const styles = theme => ({
  root: {
    display: 'flex',
    justifyContent: 'center',
    flexGrow: 1,
    marginTop: theme.spacing.unit * 30
  },
  container: {
    flexGrow: 1,
    position: 'relative',
    maxWidth: theme.spacing.unit * 75
  },
  paper: {
    position: 'absolute',
    zIndex: 1,
    marginTop: theme.spacing.unit,
    left: 0,
    right: 0
  },
  searchRoot: {
    'label + &': {
      marginTop: theme.spacing.unit * 3
    }
  },
  searchInput: {
    borderRadius: 4,
    backgroundColor: theme.palette.common.white,
    border: '1px solid #ced4da',
    fontSize: 16,
    padding: '10px 12px',
    transition: theme.transitions.create(['border-color', 'box-shadow']),
    '&:focus': {
      borderColor: '#80bdff',
      boxShadow: '0 0 0 0.2rem rgba(0,123,255,.25)'
    }
  },
  icon: {
    margin: theme.spacing.unit,
    fontSize: 250
  },
  title: {
    fontWeight: 'initial'
  }
})

function renderInput(inputProps) {
  const { InputProps, classes, ...other } = inputProps

  return (
    <Input
      classes={{
        root: classes.searchRoot,
        input: classes.searchInput
      }}
      inputProps={{
        ...InputProps
      }}
      {...other}
    />
  )
}

function renderSuggestion({
  companyName,
  index,
  itemProps,
  highlightedIndex,
  selectedItem
}) {
  const isHighlightedIndex = highlightedIndex === index
  const isSelected = (selectedItem || '').indexOf(companyName) > -1

  return (
    <MenuItem
      {...itemProps}
      key={companyName}
      selected={isHighlightedIndex}
      component="div"
      style={{
        fontWeight: isSelected ? 500 : 400
      }}
      data-testid="company-name"
    >
      {companyName}
    </MenuItem>
  )
}

renderSuggestion.propTypes = {
  highlightedIndex: PropTypes.number,
  index: PropTypes.number,
  itemProps: PropTypes.object,
  selectedItem: PropTypes.string,
  suggestion: PropTypes.shape({ label: PropTypes.string }).isRequired
}

class Search extends Component {
  render() {
    const { classes } = this.props
    const { loading, error, companiesNames } = this.props.data

    if (error) {
      return (
        <Grid container direction="column" justify="center" alignItems="center">
          <MoodBad className={classes.icon} color="disabled" />
          <Typography data-testid="graphql-error" variant="h5" color="inherit">
            Something went bad. Please, try again.
          </Typography>
        </Grid>
      )
    }

    return (
      <div className={classes.root}>
        {!loading ? (
          <div className={classes.container}>
            <Typography
              classes={{ root: classes.title }}
              component="h1"
              variant="h1"
              color="primary"
              align="center"
              gutterBottom
            >
              BigPy
            </Typography>
            <Downshift
              onChange={selectedItem => {
                this.props.history.push(`dashboard/${encodeURI(selectedItem)}`)
              }}
            >
              {({
                getInputProps,
                getMenuProps,
                getItemProps,
                highlightedIndex,
                isOpen,
                inputValue,
                selectedItem
              }) => (
                <div>
                  {renderInput({
                    fullWidth: true,
                    disableUnderline: true,
                    classes,
                    InputProps: getInputProps({
                      placeholder: 'Search a company'
                    })
                  })}
                  <div {...getMenuProps()}>
                    {isOpen ? (
                      <Paper className={classes.paper} square>
                        {companiesNames
                          .filter(name => {
                            const companyName = deburr(name.toLowerCase())
                            const searchInput = deburr(inputValue.toLowerCase())
                            return (
                              companyName.includes(searchInput) &&
                              searchInput.length > 1
                            )
                          })
                          .slice(0, MAX_SUGGESTION_ITEMS)
                          .map((companyName, index) => {
                            return renderSuggestion({
                              companyName,
                              index,
                              itemProps: getItemProps({
                                item: companyName
                              }),
                              highlightedIndex,
                              selectedItem
                            })
                          })}
                      </Paper>
                    ) : null}
                  </div>
                </div>
              )}
            </Downshift>
          </div>
        ) : (
          <BouncingLoader />
        )}
      </div>
    )
  }
}

Search.propTypes = {
  classes: PropTypes.object.isRequired
}

export const getCompaniesNamesQuery = gql`
  query {
    companiesNames
  }
`

export default withRouter(
  withStyles(styles)(graphql(getCompaniesNamesQuery)(Search))
)
