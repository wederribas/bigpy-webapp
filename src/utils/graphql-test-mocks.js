import { getCompaniesQuery } from '../components/Search/Search'

export const mocks = [
  {
    request: {
      query: getCompaniesQuery
    },
    result: {
      data: {
        allCompaniesNames: {
          edges: [
            { node: { Id: 'Company 0' } },
            { node: { Id: 'Company 1' } },
            { node: { Id: 'Company 2' } }
          ]
        }
      }
    }
  }
]
