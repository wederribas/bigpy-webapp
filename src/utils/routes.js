import Search from '../components/Search/Search'
import Top10 from '../components/Top10/Top10'
import Insights from '../components/Insights/Insights'
import About from '../components/About/About'
import Dashboard from '../components/Dashboard/Dashboard'

export const routes = [
  {
    path: '/',
    component: Search,
    exact: true
  },
  {
    path: '/dashboard/:companyName',
    component: Dashboard
  },
  {
    path: '/top10',
    component: Top10
  },
  {
    path: '/insights',
    component: Insights
  },
  {
    path: '/about',
    component: About
  }
]
