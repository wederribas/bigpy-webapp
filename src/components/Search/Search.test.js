import React from 'react'
import { Simulate } from 'react-dom/test-utils'
import { fireEvent } from 'react-testing-library'
import wait from 'waait'
import { MockedProvider } from 'react-apollo/test-utils'
import { render } from '../../utils/test-utils'
import { mocks } from '../../utils/graphql-test-mocks'
import Search from './Search'

test('Search bar filtering', async () => {
  const { container, getByTestId, getByText } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <Search />
    </MockedProvider>
  )

  expect(container.firstChild).toMatchSnapshot()

  await wait(200)

  const inputNode = container.querySelector('input')
  expect(inputNode.id).toEqual('downshift-0-input')

  Simulate.change(inputNode, { target: { value: 'Company 0' } })
  expect(getByTestId('company-name')).toHaveTextContent('Company 0')

  Simulate.change(inputNode, { target: { value: 'company' } })
  expect(getByText('Company 0')).toBeTruthy()
  expect(getByText('Company 1')).toBeTruthy()
  expect(getByText('Company 2')).toBeTruthy()

  fireEvent.click(getByText('Company 0'))
  expect(inputNode.value).toEqual('Company 0')
  expect(container.firstChild).toMatchSnapshot()
})
