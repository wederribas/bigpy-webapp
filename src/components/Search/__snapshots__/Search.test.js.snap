// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search bar filtering 1`] = `
<div
  class="Apollo-Search--root-1"
>
  <div
    class="bouncing-loader"
  >
    <div />
    <div />
    <div />
  </div>
</div>
`;

exports[`Search bar filtering 2`] = `
<div
  class="Apollo-Search--root-1"
>
  <div
    class="Apollo-Search--container-2"
  >
    <h1
      class="MuiTypography-root-8 Apollo-Search--title-7 MuiTypography-h1-20 MuiTypography-colorPrimary-38 MuiTypography-gutterBottom-35 MuiTypography-alignCenter-31"
    >
      BigPy
    </h1>
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="downshift-0-label"
      role="combobox"
    >
      <div
        class="MuiInputBase-root-56 MuiInput-root-43 Apollo-Search--searchRoot-4 MuiInputBase-fullWidth-65 MuiInput-fullWidth-50"
      >
        <input
          aria-autocomplete="list"
          aria-labelledby="downshift-0-label"
          autocomplete="off"
          class="MuiInputBase-input-66 MuiInput-input-51 Apollo-Search--searchInput-5"
          id="downshift-0-input"
          placeholder="Search a company"
          type="text"
          value="Company 0"
        />
      </div>
      <div
        aria-labelledby="downshift-0-label"
        id="downshift-0-menu"
        role="listbox"
      />
    </div>
  </div>
</div>
`;
